src/key.cpp

```diff
@@ -271,19 +271,11 @@ bool CKey::SignCompact(const uint256 &hash, std::vector<unsigned char>& vchSig)
 
 bool CKey::SignSchnorr(const uint256& hash, Span<unsigned char> sig, const uint256* merkle_root, const uint256& aux) const
 {
-    assert(sig.size() == 64);
-    bool ret = 1;
     KeyPair keypair = ComputeKeyPair(merkle_root);
     if (!keypair.IsValid()) return false;
-    ret &= secp256k1_schnorrsig_sign32(secp256k1_context_sign, sig.data(), hash.data(), reinterpret_cast<const secp256k1_keypair*>(keypair.data()), aux.data());
-    if (ret) {
-        // Additional verification step to prevent using a potentially corrupted signature
-        secp256k1_xonly_pubkey pubkey_verify;
-        ret &= secp256k1_keypair_xonly_pub(secp256k1_context_static, &pubkey_verify, nullptr, reinterpret_cast<const secp256k1_keypair*>(keypair.data()));
-        ret &= secp256k1_schnorrsig_verify(secp256k1_context_static, sig.data(), hash.begin(), 32, &pubkey_verify);
-    }
-    if (!ret) memory_cleanse(sig.data(), sig.size());
-    memory_cleanse(&keypair, sizeof(keypair));
+    bool ret = keypair.SignSchnorr(hash, sig, aux);
+    /* Clear the keypair */
+    keypair = KeyPair();
```

Comment:
Might be a naive question, but is this a behavior change, does it have any security implications (e.g. could it leave keys in the memory or something)?

Reply:
I am also wandering how the assignment involving the default constructor will clear the existing keypair data from memory, but I might be naive as well!

Reply:
> Since keypair is a local variable in this function shouldn't the KeyPair destructor do the secure erase job anyway at the end of the scope through RAII since m_keypair is a secure_unique_ptr

Correct, which should happen regardless of whether or not we explicitly clear the key data with `keypair = KeyPair();`. Having `keypair = KeyPair()` a belt and suspenders by explicitly clearing the variable. This is because if we create a `KeyPair` object with the default constructor, there is no call to `MakeKeyPairData`, hence `m_keypair` is empty. When we assign this default constructed `KeyPair` to `keypair`, it calls `ClearKeyPairData` on `keypair`, which resets it. The assignment operator was copied from `CKey` and you can see an example of the same pattern used here: https://github.com/bitcoin/bitcoin/blob/5e9f4f41de98c2cdcc74e31fe925b2f5cf9f6aec/src/bip324.cpp#L66-L70

Reply:
I'm still a bit hesitant here, see my related comment

Reply:
My last comment was wrong, @itornaza is correct: the KeyPair destructor is doing the secure erase job and the `kp = KeyPair();` is unnecessary.